nailgun-server/src/main/java/com/martiansoftware/nailgun/NGInputStream.java

```diff
@@ -18,449 +18,70 @@
-                // that we are in shutdown mode
-                throw new EOFException("stdin EOF before payload read.");
-            }
-            totalRead += currentRead;
-        }
-        return new ByteArrayInputStream(receiveBuffer);
-    }
-
-    /**
-     * Reads a NailGun chunk header from the underlying InputStream.
-     *
-     * @throws EOFException if underlying stream / socket is closed which happens on client
-     * disconnection
-     * @throws IOException if thrown by the underlying InputStream, or if an unexpected NailGun
-     * chunk type is encountered.
-     */
-    private void readChunk() throws IOException {
-        int chunkLen = din.readInt();
-        byte chunkType = din.readByte();
-        long readTime = System.currentTimeMillis();
-
-        switch (chunkType) {
-            case NGConstants.CHUNKTYPE_STDIN:
-                InputStream chunkStream = readPayload(in, chunkLen);
-                synchronized (this) {
-                    if (remaining != 0) {
-                        // TODO(buck_team) have better passthru streaming and remove this
-                        // limitation
-                        throw new IOException("Data received before stdin stream was emptied");
-                    }
-                    LOG.log(Level.FINEST, "Got stdin chunk, len {0}", chunkLen);
-                    stdin = chunkStream;
-                    remaining = chunkLen;
-                    notifyAll();
-                }
-                break;
-
-            case NGConstants.CHUNKTYPE_STDIN_EOF:
-                LOG.log(Level.FINEST, "Got stdin closed chunk");
-                setEof();
-                break;
-
-            case NGConstants.CHUNKTYPE_HEARTBEAT:
-                LOG.log(Level.FINEST, "Got client heartbeat");
-
-                ArrayList<NGHeartbeatListener> listeners;
-                synchronized (this) {
-                    // copy collection to avoid executing callbacks under lock
-                    listeners = new ArrayList<>(heartbeatListeners);
-                }
-
-                // TODO(buck_team): should probably dispatch to a different thread(pool)
-                for (NGHeartbeatListener listener : listeners) {
-                    listener.heartbeatReceived();
-                }
-
-                break;
-
-            default:
-                LOG.log(Level.WARNING, "Unknown chunk type: {0}", (char) chunkType);
-                throw new IOException("Unknown stream type: " + (char) chunkType);
-        }
-    }
-
-    /**
-     * Notify threads waiting in read() on either EOF chunk read or client disconnection.
-     */
-    private synchronized void setEof() {
-        eof = true;
-        notifyAll();
-    }
-
-    /**
-     * Notify threads waiting in read() on either EOF chunk read or client disconnection.
-     */
-    private synchronized void setClientDisconnected() {
-        clientConnected = false;
+    public NGInputStream(NGCommunicator comm){
```

Comment:
would be great to change the naming to something readable, but I understand that you're just moving things around...